### ISSUE TITLE
Cannot promote case expressions with branches that are trivially not taken

### ISSUE THREAD

**Issue by JakobBruenker** [goldfirere/singletons#425]
The following file

```Haskell
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE PolyKinds #-}

import Data.Singletons.TH

$(singletons [d|
  works x = case x of False -> ()
  doesn'tWork = case True of False -> ()
  |])
```

results in

```Haskell
    * Could not deduce: Case_6989586621681659728 TrueSym0 ~ '()
      from the context: TrueSym0 ~ 'False
        bound by a pattern with constructor: SFalse :: SBool 'False,
                 in a case alternative
        at MyTest.hs:(12,3)-(15,4)
      Expected type: Sing
                       (Case_6989586621681659728
                          Let6989586621681659726Scrutinee_6989586621681659723Sym0)
        Actual type: STuple0 '()
    * In the expression: STuple0
      In a case alternative: SFalse -> STuple0
      In the second argument of `id', namely
        `(case sScrutinee_6989586621681659723 of { SFalse -> STuple0 })'
   |
12 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...
```

with HEAD, if `doesn'tWork` is not commented out.

This might not be terribly high-priority to fix though, since I can't really think of a reason why one would need this.

**Comment by RyanGlScott** [comment]
I don't see a way to make this code typecheck due to how `singletons` encodes pattern matching. You're basically writing this code:

```hs
type family Doesn'tWork :: () where
  Doesn'tWork = Case True

type family Case x where
  Case False = '()

sDoesn'tWork :: Sing Doesn'tWork
sDoesn'tWork =
  case STrue of
    SFalse -> STuple0
```
```
    • Could not deduce: Case 'True ~ '()
      from the context: 'True ~ 'False
        bound by a pattern with constructor: SFalse :: SBool 'False,
                 in a case alternative
        at Bug.hs:21:5-10
      Expected type: Sing Doesn'tWork
        Actual type: STuple0 '()
    • In the expression: STuple0
      In a case alternative: SFalse -> STuple0
      In the expression: case STrue of { SFalse -> STuple0 }
```

One thing that's interesting to note here is that the `SFalse` case is completely unreachable. In fact, if you were to replace `STuple0` with, say, `undefined`, GHC is helpful enough to tell you this:

```
Bug.hs:21:5: warning: [-Winaccessible-code]
    • Couldn't match type ‘'True’ with ‘'False’
      Inaccessible code in
        a pattern with constructor: SFalse :: SBool 'False,
        in a case alternative
    • In the pattern: SFalse
      In a case alternative: SFalse -> undefined
      In the expression: case STrue of { SFalse -> undefined }
   |
21 |     SFalse -> undefined
   |     ^^^^^^

Bug.hs:21:5: warning: [-Woverlapping-patterns]
    Pattern match has inaccessible right hand side
    In a case alternative: SFalse -> ...
   |
21 |     SFalse -> undefined
   |     ^^^^^^^^^^^^^^^^^^^
```

So we are already walking on shaky ground. The only reason that using `STuple0` as the right-hand side causes a full-blown type error is due to a quirk in how GHC works, it will attempt to typecheck inaccessible case alternatives even though they're, well, inaccessible. In other words, it will proceed under the false premise that `True ~ False` and see what happens. (This is why the error message mentions "`from the context: 'True ~ 'False`".)

As for why this gives a type error in the first place, note that the return type is `Sing Doesn'tWork`, which is equivalent to `Sing (Case True)`. But `Case True` is a stuck type, which means that it does not reduce to `'()`, resulting in the "`Could not deduce: Case 'True ~ '()`" error you see above. While you can get away with this sort of trickery in the original, un-singled code, singling it imbues the code with richer type information that causes the house of cards to topple over (See also the "overlapping patterns" section of [the `singletons` README](https://github.com/goldfirere/singletons/blob/07f4422246ca4d37adf523b25b39346b2a3199b3/README.md) for another example of pattern-matching that is not easily possible in singled code.)

In short, I don't think this is feasible. Do you feel otherwise?

**Comment by JakobBruenker** [comment]
Thank you for the explanation.

> In short, I don't think this is feasible.

That seems like a fair assessment. Feel free to close.
